bionic/persistence.py

```diff
@@ -149,27 +149,50 @@ def load_artifact(self):
         Returns the nearest cached artifact for this provenance, if one exists.
         """
 
+        try:
+            entry = self._get_nearest_entry_with_artifact()
+
+            if entry is None:
+                return None
+
+            return entry.artifact
+
+        except InternalCacheStateError as e:
+            self.raise_state_error_with_explanation(e)
+
+    def load_local_artifact(self):
+        """
+        Returns a local artifact if any cached artifact exists.
+
+        If the nearest artifact is not local, it will be copied to a local instance and
+        registered in the cache.
+        """
+
         try:
             entry = self._get_nearest_entry_with_artifact()
 
             if entry is None:
                 return None
 
             if entry.tier == "local":
-                return entry.artifact
+                local_artifact = entry.artifact
 
             elif entry.tier == "cloud":
-                return self._local_artifact_from_cloud(entry.artifact)
+                local_artifact = self._local_artifact_from_cloud(entry.artifact)
 
             else:
                 raise AssertionError("Unrecognized tier: " + entry.tier)
 
+            self._save_or_reregister_artifact(local_artifact)
```

Comment:
I was wondering about doing this also. I was thinking if this can cause unnecessary uploads to GCS. For example, we try to load an artifact from a previous computation; the new code would cause the disk copy to be uploaded to GCS if it is not already there; and it later turns out that we need an exact match while the artifact only has a functional match. Then, we will need to re-compute the entity anyways.

Reply:
Good question. Can this happen? I haven't thought this all the way through, but we only call `load_local_artifact` when we're ready to load the data into memory, so I think the versioning check for this task should have already happened.

Also, two other things have to be true for this to happen:

1. The artifact wasn't already in GCS, which means we computed it with GCS caching off and then ran the flow again with GCS caching on. That can definitely happen, but probably not very often.
2. The user made a versioning mistake.

Given all these qualifications, this change seems pretty safe to me.